Logging with build tags

it1804 opened the pull request:
A possible solution to the logging performance issue is by using build tags.

Latest version 1.31
go test -v -bench=. -benchmem -benchtime=100000x -tags all
```
BenchmarkSendOneRequest-24        100000             80601 ns/op            3259 B/op         90 allocs/op
```
New Version based on 1.31 with logging enabled 
go test -v -bench=. -benchmem -benchtime=100000x -tags all
```
BenchmarkSendOneRequest-24        100000             69803 ns/op            3088 B/op         84 allocs/op
```
New Version based on 1.31 with logging disabled
go test -v -bench=. -benchmem -benchtime=100000x -tags all,gosnmp_nodebug
```
BenchmarkSendOneRequest-24        100000             65647 ns/op            2592 B/op         53 allocs/op
```

The interface for configuring logging has been changed.
Before:
`Logger: log.New (os.Stdout, "", 0)`
After:
`Logger: NewLogger (log.New (os.Stdout, "", 0) `

Logging can be completely disabled using the "gosnmp_nodebug" build tag (go build -tags gosnmp_nodebug), even if the logger has been installed.

Also, the interface was unified to a single style, instead of the previously used logPrint(f) and Print(f) 

Now there is no need to use ioutil.Discard

SuperQ commented on logger.go:
```diff
@@ -0,0 +1,9 @@
+// +build !gosnmp_debug
```
This needs to be flipped around. This is a relatively small optimization, and most users will expect this to work by default.

```suggestion
// +build nogosnmp_debug
```

it1804 replied:
Replaced with gosnmp_nodebug

SuperQ commented on logger_debug.go:
```diff
@@ -0,0 +1,15 @@
+// +build gosnmp_debug
```
```suggestion
// +build !nogosnmp_debug
```

TimRots commented on gosnmp.go:
```diff
@@ -71,20 +69,16 @@ type GoSNMP struct {
 
-	// loggingEnabled is set if the Logger isn't nil, otherwise any logging calls
-	// are ignored via shortcircuit.
-	loggingEnabled bool
```
The deprecation of this flag needs to be mentioned in the unreleased ``NOTE`` section in ``CHANGELOG.md``.
Could you as well add a ``[CHANGE]:`` entry for this pr?

it1804 replied:
Added a description. How complete do you think it is?

TimRots commented on logger.go:
```diff
@@ -0,0 +1,9 @@
+// +build gosnmp_nodebug
+
```
Please also add the instructions on how to disable the logger via a build tag here

it1804 replied:
Fixed

TimRots reviewed (commented):
Left some minor nits for documenting these changes in the ``CHANGELOG`` and ``README``. 
Thank you @it1804

it1804 commented on trap_test.go:
```diff
@@ -33,6 +33,7 @@ const (
 	trapTestTimestamp     = 300
 )
 
+/*
```
I'm sorry. This was erroneously commented out during debugging. I'll fix it.

it1804 replied:
I checked again and it turned out that it was not my fault. In fact, this structure has never been used by anyone.
A test that was supposed to use this structure (TestUnmarshalTrap ()) was broken and commented out earlier.
I have fixed this test, but you will need to check if it is correct when I update the PR.

it1804 commented on trap_test.go:
```diff
@@ -52,6 +53,7 @@ var testsUnmarshalTrap = []struct {
 		},
 	},
 }
+*/
```
This was also erroneously commented out. I'll fix it.

it1804 replied:
Answer above

it1804 commented:
Updated documentation and fixed TestUnmarshalTrap() test. Please check if it is correct.

PS. There was a conflict with https://github.com/gosnmp/gosnmp/pull/331 If you accept this PR, the check added in it will not be needed.

it1804 commented on gosnmp.go:
```diff
@@ -71,20 +69,16 @@ type GoSNMP struct {
+	// Logger is the GoSNMP.Logger to use for debugging.
+	// For verbose logging to stdout:
+	// x.Logger = NewLogger(log.New(os.Stdout, "", 0))
+	// For Release builds, you can turn off logging entirely by using the go build tag "gosnmp_nodebug" even if the logger was installed.
```
Added a description. How complete do you think it is?

it1804 commented on helper.go:
```diff
@@ -623,9 +623,7 @@ func parseRawField(logger Logger, data []byte, msg string) (interface{}, int, er
 	if len(data) == 0 {
 		return nil, 0, fmt.Errorf("empty data passed to parseRawField")
 	}
-	if logger != nil {
```
This check is no longer needed

TimRots reviewed (approved):
LGTM

AlexandreYang reviewed (commented):
Thanks a lot for this PR !

If I understand correctly, when when logger is enabled with build tags `+build !gosnmp_nodebug`, the `gosnmp.Default.Logger/gosnmp.GoSNMP.Logger` will default to `nil` and no logs will be printed due to `if` condition in `Logger.Print/Printf` ?

Asking that because we have a use where we need to enabled/disable gosnmp logs at runtime.

--

Also, could be interesting to also have some benchmark values for the case "New Version based on 1.31 with logging enabled via build tags + logger set to nil"

SuperQ commented:
@AlexandreYang If you're looking to enable/disable at runtime, you don't build with the tag.

The build tag is only needed for people who never want to have logs and want to have the compiler optimize the whole function call out of the code. This is a small performance improvement for those users.

it1804 commented:
> 
> 
> @it1804 Thanks a lot for this PR !
> 
> If I understand correctly, when when logger is enabled with build tags `+build !gosnmp_nodebug`, the `gosnmp.Default.Logger/gosnmp.GoSNMP.Logger` will default to `nil` and no logs will be printed due to `if` condition in `Logger.Print/Printf` ?
> 
> Asking that because we have a use where we need to enabled/disable gosnmp logs at runtime.
> 
> --
> 
> Also, could be interesting to also have some benchmark values for the case "New Version based on 1.31 with logging enabled via build tags + logger set to nil"

If you want to enable / disable logging at runtime, building with the gosnmp_nodebug tag will not work for you. Build with this tag will completely disable compiler-level logging, slightly optimizing performance.

You can switch logging at runtime, for example, like this:
```go
package main

import (
	g "github.com/gosnmp/gosnmp"
)

func main() {

	stdout_logger: = NewLogger(log.New(os.Stdout, "", 0)) // enable logging with stdout
	disabled_logger: = NewLogger(nil)  // disable logging

	params := &g.GoSNMP{
		Target:    "127.0.0.1",
		Port:      uint16(1161),
		Community: "public",
	}
	params.Connect()  // no logger specified, logging is disabled 
	params.Conn.Close()

	params.Logger = stdout_logger
	params.Connect() // logging enabled using stdout 
	params.Conn.Close()
	
	params.Logger = disabled_logger
	params.Connect() // logging is disabled
	params.Conn.Close()
}
```
My test shows no difference between an uninitialized Logger variable and a value set to zero. 

New Version based on 1.31 with logging enabled, and Logger variable is not set
go test -v -bench=. -benchmem -benchtime=100000x -tags all
```
BenchmarkSendOneRequest-24        100000             70542 ns/op            3088 B/op         84 allocs/op
```

New Version based on 1.31 with logging enabled, and Logger variable is set to NewLogger(nil)
go test -v -bench=. -benchmem -benchtime=100000x -tags all
```
BenchmarkSendOneRequest-24        100000             70377 ns/op            3088 B/op         84 allocs/op
```

TimRots commented:
@SuperQ Could you give this PR a final review please.

AlexandreYang commented:
Just notice this might be a breaking change if `g.Default.Logger = log.New(os.Stdout, "", 0)` is currently used. It will lead to:
> cannot use log.New(os.Stdout, "", 0) (type *log.Logger) as type gosnmp.Logger in assignment

It quite easy to migrate, but just wondering if we are fine with that ?

it1804 commented:
> 
> 
> Just noticed this might be a breaking change (if I'm not missing something) if `g.Default.Logger = log.New(...)` is currently used. It will lead to:
> 
> > cannot use log.New(os.Stdout, "", 0) (type *log.Logger) as type gosnmp.Logger in assignment
> 
> It quite easy to migrate, but just wondering if we are fine with that ?
> Might worth trying to keep backward compatibility ?

Not sure if I can do backward compatibility. The purpose of this PR was to get rid of the pointer in the Logger variable, i.e. log.New (...) is the pointer in the current implementation.
If you do this again, you may again have a problem with nil when passed to helper functions, which again requires you to go back to using ioutil.Discard and some other checks. I think then we can go back to the situation described at https://github.com/gosnmp/gosnmp/issues/313.
The main result of the changes made here is to move the external registrar interface inside the internal registrar object to abstract the external registrar functions from the internal gosnmp interfaces (functions).

SuperQ commented:
I'm OK with the breaking change. There's a nice small improvement for both enabled and disabled logging. It also eliminates the need to check `if logger != nil`, which is very nice for safety.

SuperQ commented on README.md:
````diff
@@ -48,15 +48,33 @@ will require modification in these (and other) locations:
+To enable logging, you must call gosnmp.NewLogger() function, and pass a pointer to your logging interface, for example with standard *log.Logger:
+
+```go
+gosnmp.Default.Logger = NewLogger(log.New(os.Stdout, "", 0))
````
Shouldn't this be `gosnmp.NewLogger()`?

it1804 replied:
Fixed

SuperQ commented on README.md:
````diff
@@ -48,15 +48,33 @@ will require modification in these (and other) locations:
+```go
+g := &gosnmp.GoSNMP{
+    ...
+    Logger:    NewLogger(log.New(os.Stdout, "", 0)),
````
Same here, `gosnmp.NewLogger()`?

it1804 replied:
Fixed

SuperQ commented on README.md:
````diff
@@ -48,15 +48,33 @@ will require modification in these (and other) locations:
+```
+go build -tags gosnmp_nodebug
+```
+This will completely disable the logging of the gosnmp library, even if the logger interface is specified in the code. This will allow you to manage debug / release builds without changing the code, and will also improve performance. Whenever possible, always create release builds using this tag.
````
I don't agree with this recommendation. Logging is important for production use. Without good logs, diagnosing production problems is made more difficult.

```suggestion
This will completely disable the logging of the gosnmp library, even if the logger interface is specified in the code. This provides a small performance improvement.```

it1804 replied:
I agree, I wanted to change that too. My original formulation looks ambiguous. Fixed.

SuperQ reviewed (approved):
LGTM, thanks a bunch for working on this.

SuperQ commented:
I think the only thing left is to rebase to fix the conflict.
